src/error.rs

```diff
@@ -1,39 +1,124 @@
 #![allow(overflowing_literals)]
 
+use crate::runtime::*;
+use crate::*;
+
+/// The ErrorCode (a.k.a HRESULT) of an error
+#[repr(transparent)]
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub struct ErrorCode(pub i32);
+
+/// A WinRT related error
+pub struct Error {
+    code: ErrorCode,
+    info: IRestrictedErrorInfo,
+}
+
 /// An alias for `std::result::Result<T, winrt::Error>`
 #[must_use]
 pub type Result<T> = std::result::Result<T, Error>;
 
 impl<T> std::convert::From<Result<T>> for ErrorCode {
     fn from(result: Result<T>) -> Self {
         if let Err(error) = result {
-            // TODO: call SetErrorInfo
+            if !error.info.is_null() {
+                unsafe {
+                    SetRestrictedErrorInfo(error.info.as_raw() as _);
+                }
+            }
             return error.code();
         }
 
         ErrorCode(0)
     }
 }
 
-/// A WinRT related error
-#[derive(Debug)]
-pub struct Error {
-    code: ErrorCode,
-    // TODO: add `info: IErrorInfo`
+impl std::convert::From<ErrorCode> for Error {
+    fn from(code: ErrorCode) -> Self {
+        unsafe {
+            let mut info = IErrorInfo::default();
+            GetErrorInfo(0, info.set_abi() as _);
+            let restricted = info.query::<IRestrictedErrorInfo>();
+
+            if !restricted.is_null() {
+                let info2 = info.query::<ILanguageExceptionErrorInfo2>();
```

Comment:
nit: I think just shadowing the `info` binding is better than using a name like `info2`

Reply:
I used it to clearly identify that it is referring to a `ILanguageExceptionErrorInfo2` (note the `Info2` suffix) as apposed to one of the regular info objects. These interfaces don't have great names. Perhaps I'll call this variable `capture` to make it clearer what this is about.

Reply:
OH... sorry. Yes weird name... But that makes the variable name better considering...